### ISSUE TITLE
[no-unnecessary-condition] False positive when infer type from class methods

### ISSUE THREAD

**Issue by chengB12** [typescript-eslint/typescript-eslint#2424]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.
**Repo**
```ts
{
  "rules": {
    '@typescript-eslint/no-unnecessary-condition': ['error', {
            allowConstantLoopConditions: true
      }],
  }
}
```

```TS
class Client {
    getValue(): number | undefined {
        return 1
    }
}
const client = new Client()
export class Foo {
    bar() {
        const r = client.getValue()
        if (!r) {
            //
        }
    }
}
```
tsconfig.json
```json
{
    "compilerOptions": {
      /* Basic Options */
      "target": "es2019",
      "module": "commonjs",
      "sourceMap": true,
      "jsx": "react",
      "resolveJsonModule": true,
      "strict": true,
      "strictNullChecks": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "noImplicitReturns": true,
      "noFallthroughCasesInSwitch": true,
      "esModuleInterop": true,
      "experimentalDecorators": true,
      "emitDecoratorMetadata": true,
    },
}

```
**Expected Result**

No alert at all

**Actual Result**
```JSON
{
	"resource": "/c:/work/vdm-services/packages/testdata/helper-scripts/foo.ts",
	"owner": "eslint",
	"code": {
		"value": "@typescript-eslint/no-unnecessary-condition",
		"target": {
			"$mid": 1,
			"external": "https://github.com/typescript-eslint/typescript-eslint/blob/v3.10.0/packages/eslint-plugin/docs/rules/no-unnecessary-condition.md",
			"path": "/typescript-eslint/typescript-eslint/blob/v3.10.0/packages/eslint-plugin/docs/rules/no-unnecessary-condition.md",
			"scheme": "https",
			"authority": "github.com"
		}
	},
	"severity": 8,
	"message": "Unnecessary conditional, value is always falsy.",
	"source": "eslint",
	"startLineNumber": 10,
	"startColumn": 13,
	"endLineNumber": 10,
	"endColumn": 15
}
```
Debug log:
```
$ C:\work\vdm-services\node_modules\.bin\eslint packages\testdata\helper-scripts\foo.ts --debug
  eslint:cli CLI args: [ 'packages\\testdata\\helper-scripts\\foo.ts', '--debug' ] +0ms
  eslint:cli Running on files +3ms
  eslint:config-array-factory Loading JSON config file: C:\work\vdm-services\package.json +0ms
  eslint:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: 'C:\\work\\vdm-services', loose: false } ] +0ms
  eslint:ignore-pattern   processed: { basePath: 'C:\\work\\vdm-services', patterns: [ '/**/node_modules/*' ] } +3ms
  eslint:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: 'C:\\work\\vdm-services', loose: false } ] +0ms
  eslint:ignore-pattern   processed: { basePath: 'C:\\work\\vdm-services', patterns: [ '/**/node_modules/*' ] } +1ms
  eslint:file-enumerator Start to iterate files: [ 'packages\\testdata\\helper-scripts\\foo.ts' ] +0ms
  eslint:file-enumerator File: C:\work\vdm-services\packages\testdata\helper-scripts\foo.ts +19ms
  eslint:cascading-config-array-factory Load config files for C:\work\vdm-services\packages\testdata\helper-scripts. +0ms
  eslint:cascading-config-array-factory No cache found: C:\work\vdm-services\packages\testdata\helper-scripts. +2ms
  eslint:config-array-factory Config file not found on C:\work\vdm-services\packages\testdata\helper-scripts +35ms
  eslint:cascading-config-array-factory No cache found: C:\work\vdm-services\packages\testdata. +2ms
  eslint:config-array-factory Loading package.json config file: C:\work\vdm-services\packages\testdata\package.json +20ms
  eslint:config-array-factory Loading JSON config file: C:\work\vdm-services\packages\testdata\package.json +1ms
  eslint:config-array-factory Error reading package.json file: C:\work\vdm-services\packages\testdata\package.json +1ms
  eslint:config-array-factory Config file not found on C:\work\vdm-services\packages\testdata +0ms
  eslint:cascading-config-array-factory No cache found: C:\work\vdm-services\packages. +22ms
  eslint:config-array-factory Config file not found on C:\work\vdm-services\packages +1ms
  eslint:cascading-config-array-factory No cache found: C:\work\vdm-services. +0ms
  eslint:config-array-factory Loading JS config file: C:\work\vdm-services\.eslintrc.js +1ms
  eslint:config-array-factory Config file found: C:\work\vdm-services\.eslintrc.js +2ms
  eslint:config-array-factory Loading {extends:"plugin:@typescript-eslint/recommended"} relative to C:\work\vdm-services\.eslintrc.js +1ms
  eslint:config-array-factory Loading plugin "@typescript-eslint" from C:\work\vdm-services\.eslintrc.js +1ms
  eslint:config-array-factory Loaded: @typescript-eslint/eslint-plugin@3.10.0 (C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\index.js) +4ms
  eslint:config-array-factory Plugin C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\index.js loaded in: 324ms +348ms
  eslint:config-array-factory Loading {extends:"./configs/base"} relative to C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\index.js +1ms
  eslint:config-array-factory package.json was not found: Cannot find module './configs/base/package.json'
Require stack:
- C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\index.js +1ms
  eslint:config-array-factory Loaded: ./configs/base (C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\configs\base.js) +23ms
  eslint:config-array-factory Loading JS config file: C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\configs\base.js +0ms
  eslint:config-array-factory Loading parser "@typescript-eslint/parser" from C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\configs\base.js +1ms
  eslint:config-array-factory Loaded: @typescript-eslint/parser@3.10.0 (C:\work\vdm-services\node_modules\@typescript-eslint\parser\dist\index.js) +2ms
  eslint:config-array-factory Loading plugin "@typescript-eslint" from C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\configs\base.js +71ms
  eslint:config-array-factory Loaded: @typescript-eslint/eslint-plugin@3.10.0 (C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\index.js) +1ms
  eslint:config-array-factory Plugin C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\index.js loaded in: 0ms +0ms
  eslint:config-array-factory Loading {extends:"./configs/eslint-recommended"} relative to C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\index.js +0ms
  eslint:config-array-factory package.json was not found: Cannot find module './configs/eslint-recommended/package.json'
Require stack:
- C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\index.js +1ms
  eslint:config-array-factory Loaded: ./configs/eslint-recommended (C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\configs\eslint-recommended.js) +0ms
  eslint:config-array-factory Loading JS config file: C:\work\vdm-services\node_modules\@typescript-eslint\eslint-plugin\dist\configs\eslint-recommended.js +1ms
  eslint:config-array-factory Loading {extends:"eslint:recommended"} relative to C:\work\vdm-services\.eslintrc.js +2ms
  eslint:config-array-factory Loading JS config file: C:\work\vdm-services\node_modules\eslint\conf\eslint-recommended.js +0ms
  eslint:config-array-factory Loading parser "@typescript-eslint/parser" from C:\work\vdm-services\.eslintrc.js +1ms
  eslint:config-array-factory Loaded: @typescript-eslint/parser@3.10.0 (C:\work\vdm-services\node_modules\@typescript-eslint\parser\dist\index.js) +1ms
  eslint:cascading-config-array-factory No cache found: C:\work. +463ms
  eslint:config-array-factory Config file not found on C:\work +1ms
  eslint:cascading-config-array-factory No cache found: C:\. +1ms
  eslint:config-array-factory Config file not found on C:\ +1ms
  eslint:rules Loading rule 'constructor-super' (remaining=281) +0ms
  eslint:rules Loading rule 'getter-return' (remaining=280) +1ms
  eslint:rules Loading rule 'no-const-assign' (remaining=279) +1ms
  eslint:rules Loading rule 'no-dupe-args' (remaining=278) +0ms
  eslint:rules Loading rule 'no-dupe-class-members' (remaining=277) +1ms
  eslint:rules Loading rule 'no-dupe-keys' (remaining=276) +0ms
  eslint:rules Loading rule 'no-func-assign' (remaining=275) +1ms
  eslint:rules Loading rule 'no-import-assign' (remaining=274) +1ms
  eslint:rules Loading rule 'no-new-symbol' (remaining=273) +23ms
  eslint:rules Loading rule 'no-obj-calls' (remaining=272) +1ms
  eslint:rules Loading rule 'no-redeclare' (remaining=271) +0ms
  eslint:rules Loading rule 'no-setter-return' (remaining=270) +1ms
  eslint:rules Loading rule 'no-this-before-super' (remaining=269) +1ms
  eslint:rules Loading rule 'no-undef' (remaining=268) +1ms
  eslint:rules Loading rule 'no-unreachable' (remaining=267) +1ms
  eslint:rules Loading rule 'no-unsafe-negation' (remaining=266) +0ms
  eslint:rules Loading rule 'no-var' (remaining=265) +1ms
  eslint:rules Loading rule 'prefer-const' (remaining=264) +2ms
  eslint:rules Loading rule 'prefer-rest-params' (remaining=263) +2ms
  eslint:rules Loading rule 'prefer-spread' (remaining=262) +1ms
  eslint:rules Loading rule 'valid-typeof' (remaining=261) +21ms
  eslint:rules Loading rule 'no-array-constructor' (remaining=260) +5ms
  eslint:rules Loading rule 'no-empty-function' (remaining=259) +1ms
  eslint:rules Loading rule 'no-extra-semi' (remaining=258) +2ms
  eslint:rules Loading rule 'no-unused-vars' (remaining=257) +2ms
  eslint:rules Loading rule 'for-direction' (remaining=256) +4ms
  eslint:rules Loading rule 'no-async-promise-executor' (remaining=255) +17ms
  eslint:rules Loading rule 'no-case-declarations' (remaining=254) +1ms
  eslint:rules Loading rule 'no-class-assign' (remaining=253) +1ms
  eslint:rules Loading rule 'no-compare-neg-zero' (remaining=252) +1ms
  eslint:rules Loading rule 'no-cond-assign' (remaining=251) +0ms
  eslint:rules Loading rule 'no-constant-condition' (remaining=250) +2ms
  eslint:rules Loading rule 'no-control-regex' (remaining=249) +1ms
  eslint:rules Loading rule 'no-debugger' (remaining=248) +1ms
  eslint:rules Loading rule 'no-delete-var' (remaining=247) +1ms
  eslint:rules Loading rule 'no-dupe-else-if' (remaining=246) +1ms
  eslint:rules Loading rule 'no-duplicate-case' (remaining=245) +1ms
  eslint:rules Loading rule 'no-empty' (remaining=244) +1ms
  eslint:rules Loading rule 'no-empty-character-class' (remaining=243) +1ms
  eslint:rules Loading rule 'no-empty-pattern' (remaining=242) +1ms
  eslint:rules Loading rule 'no-ex-assign' (remaining=241) +26ms
  eslint:rules Loading rule 'no-extra-boolean-cast' (remaining=240) +1ms
  eslint:rules Loading rule 'no-fallthrough' (remaining=239) +1ms
  eslint:rules Loading rule 'no-global-assign' (remaining=238) +2ms
  eslint:rules Loading rule 'no-inner-declarations' (remaining=237) +2ms
  eslint:rules Loading rule 'no-invalid-regexp' (remaining=236) +1ms
  eslint:rules Loading rule 'no-irregular-whitespace' (remaining=235) +2ms
  eslint:rules Loading rule 'no-misleading-character-class' (remaining=234) +17ms
  eslint:rules Loading rule 'no-mixed-spaces-and-tabs' (remaining=233) +19ms
  eslint:rules Loading rule 'no-octal' (remaining=232) +1ms
  eslint:rules Loading rule 'no-prototype-builtins' (remaining=231) +1ms
  eslint:rules Loading rule 'no-regex-spaces' (remaining=230) +1ms
  eslint:rules Loading rule 'no-self-assign' (remaining=229) +1ms
  eslint:rules Loading rule 'no-shadow-restricted-names' (remaining=228) +1ms
  eslint:rules Loading rule 'no-sparse-arrays' (remaining=227) +1ms
  eslint:rules Loading rule 'no-unexpected-multiline' (remaining=226) +2ms
  eslint:rules Loading rule 'no-unsafe-finally' (remaining=225) +1ms
  eslint:rules Loading rule 'no-unused-labels' (remaining=224) +1ms
  eslint:rules Loading rule 'no-useless-catch' (remaining=223) +1ms
  eslint:rules Loading rule 'no-useless-escape' (remaining=222) +1ms
  eslint:rules Loading rule 'no-with' (remaining=221) +1ms
  eslint:rules Loading rule 'require-yield' (remaining=220) +1ms
  eslint:rules Loading rule 'use-isnan' (remaining=219) +16ms
  eslint:rules Loading rule 'no-console' (remaining=218) +3ms
  eslint:rules Loading rule 'eqeqeq' (remaining=217) +1ms
  eslint:rules Loading rule 'max-classes-per-file' (remaining=216) +2ms
  eslint:rules Loading rule 'max-len' (remaining=215) +1ms
  eslint:rules Loading rule 'semi' (remaining=214) +5ms
  eslint:rules Loading rule 'quote-props' (remaining=213) +1ms
  eslint:rules Loading rule 'quotes' (remaining=212) +2ms
  eslint:rules Loading rule 'new-parens' (remaining=211) +2ms
  eslint:rules Loading rule 'no-caller' (remaining=210) +18ms
  eslint:rules Loading rule 'no-bitwise' (remaining=209) +13ms
  eslint:rules Loading rule 'indent' (remaining=208) +2ms
  eslint:rules Loading rule 'no-floating-decimal' (remaining=207) +26ms
  eslint:rules Loading rule 'comma-dangle' (remaining=206) +1ms
  eslint:rules Loading rule 'comma-spacing' (remaining=205) +2ms
  eslint:rules Loading rule 'no-trailing-spaces' (remaining=204) +1ms
  eslint:rules Loading rule 'curly' (remaining=203) +2ms
  eslint:rules Loading rule 'prefer-arrow-callback' (remaining=202) +3ms
  eslint:rules Loading rule 'object-shorthand' (remaining=201) +14ms
  eslint:rules Loading rule 'multiline-ternary' (remaining=200) +19ms
  eslint:rules Loading rule 'operator-linebreak' (remaining=199) +14ms
  eslint:rules Loading rule 'no-promise-executor-return' (remaining=198) +3ms
  eslint:rules Loading rule 'no-template-curly-in-string' (remaining=197) +24ms
  eslint:rules Loading rule 'no-throw-literal' (remaining=196) +8ms
  eslint:rules Loading rule 'no-useless-concat' (remaining=195) +2ms
  eslint:rules Loading rule 'no-useless-return' (remaining=194) +1ms
  eslint:rules Loading rule 'no-shadow' (remaining=193) +1ms
  eslint:rules Loading rule 'no-restricted-imports' (remaining=192) +59ms
  eslint:rules Loading rule 'no-unused-expressions' (remaining=191) +4ms
  eslint:cascading-config-array-factory Configuration was determined: ConfigArray(9) [ { type: 'config', name: 'DefaultIgnorePattern', filePath: '', criteria: null, env: undefined, globals: undefined, ignorePattern: IgnorePattern { patterns: [Array], basePath: 'C:\\work\\vdm-services', loose: false }, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined }, { type: 'config', name: '.eslintrc.js » plugin:@typescript-eslint/recommended » ./configs/base', filePath: 'C:\\work\\vdm-services\\node_modules\\@typescript-eslint\\eslint-plugin\\dist\\configs\\base.js', criteria: null, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: { error: null, filePath: 'C:\\work\\vdm-services\\node_modules\\@typescript-eslint\\parser\\dist\\index.js', id: '@typescript-eslint/parser', importerName: '.eslintrc.js » plugin:@typescript-eslint/recommended » ./configs/base', importerPath: 'C:\\work\\vdm-services\\node_modules\\@typescript-eslint\\eslint-plugin\\dist\\configs\\base.js' }, parserOptions: { sourceType: 'module' }, plugins: { '@typescript-eslint': [Object] }, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined }, { type: 'config', name: '.eslintrc.js » plugin:@typescript-eslint/recommended » ./configs/eslint-recommended', filePath: 'C:\\work\\vdm-services\\node_modules\\@typescript-eslint\\eslint-plugin\\dist\\configs\\eslint-recommended.js', criteria: null, env: undefined, globals: undefined, 
ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined }, { type: 'config', name: '.eslintrc.js » plugin:@typescript-eslint/recommended » ./configs/eslint-recommended#overrides[0]', filePath: 'C:\\work\\vdm-services\\node_modules\\@typescript-eslint\\eslint-plugin\\dist\\configs\\eslint-recommended.js', criteria: { includes: [Array], excludes: null, basePath: 'C:\\work\\vdm-services' }, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { 'constructor-super': 
'off', 'getter-return': 'off', 'no-const-assign': 'off', 'no-dupe-args': 'off', 'no-dupe-class-members': 'off', 'no-dupe-keys': 'off', 'no-func-assign': 'off', 'no-import-assign': 'off', 'no-new-symbol': 'off', 'no-obj-calls': 'off', 'no-redeclare': 'off', 'no-setter-return': 'off', 'no-this-before-super': 'off', 'no-undef': 'off', 'no-unreachable': 'off', 'no-unsafe-negation': 'off', 'no-var': 'error', 'prefer-const': 'error', 'prefer-rest-params': 'error', 'prefer-spread': 'error', 'valid-typeof': 'off' }, settings: undefined }, { type: 'config', name: '.eslintrc.js » plugin:@typescript-eslint/recommended', filePath: 'C:\\work\\vdm-services\\node_modules\\@typescript-eslint\\eslint-plugin\\dist\\index.js', criteria: null, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { '@typescript-eslint/adjacent-overload-signatures': 'error', '@typescript-eslint/ban-ts-comment': 'error', '@typescript-eslint/ban-types': 
'error', '@typescript-eslint/explicit-module-boundary-types': 'warn', 'no-array-constructor': 'off', '@typescript-eslint/no-array-constructor': 'error', 'no-empty-function': 'off', '@typescript-eslint/no-empty-function': 'error', '@typescript-eslint/no-empty-interface': 'error', '@typescript-eslint/no-explicit-any': 'warn', '@typescript-eslint/no-extra-non-null-assertion': 'error', 'no-extra-semi': 'off', '@typescript-eslint/no-extra-semi': 'error', '@typescript-eslint/no-inferrable-types': 'error', '@typescript-eslint/no-misused-new': 'error', '@typescript-eslint/no-namespace': 'error', '@typescript-eslint/no-non-null-asserted-optional-chain': 'error', '@typescript-eslint/no-non-null-assertion': 'warn', '@typescript-eslint/no-this-alias': 'error', 'no-unused-vars': 'off', '@typescript-eslint/no-unused-vars': 'warn', '@typescript-eslint/no-var-requires': 'error', '@typescript-eslint/prefer-as-const': 'error', '@typescript-eslint/prefer-namespace-keyword': 'error', '@typescript-eslint/triple-slash-reference': 'error' }, settings: undefined }, { type: 'config', name: '.eslintrc.js » eslint:recommended', filePath: 'C:\\work\\vdm-services\\node_modules\\eslint\\conf\\eslint-recommended.js', criteria: null, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { 'constructor-super': 'error', 'for-direction': 'error', 'getter-return': 'error', 'no-async-promise-executor': 'error', 'no-case-declarations': 'error', 'no-class-assign': 'error', 'no-compare-neg-zero': 'error', 'no-cond-assign': 'error', 'no-const-assign': 'error', 'no-constant-condition': 'error', 'no-control-regex': 'error', 'no-debugger': 'error', 'no-delete-var': 'error', 'no-dupe-args': 'error', 'no-dupe-class-members': 'error', 'no-dupe-else-if': 'error', 'no-dupe-keys': 'error', 'no-duplicate-case': 'error', 'no-empty': 'error', 'no-empty-character-class': 'error', 'no-empty-pattern': 'error', 'no-ex-assign': 'error', 'no-extra-boolean-cast': 'error', 'no-extra-semi': 'error', 'no-fallthrough': 'error', 'no-func-assign': 'error', 'no-global-assign': 'error', 'no-import-assign': 'error', 'no-inner-declarations': 'error', 'no-invalid-regexp': 'error', 'no-irregular-whitespace': 'error', 'no-misleading-character-class': 'error', 'no-mixed-spaces-and-tabs': 'error', 'no-new-symbol': 'error', 'no-obj-calls': 'error', 'no-octal': 'error', 'no-prototype-builtins': 'error', 'no-redeclare': 'error', 'no-regex-spaces': 'error', 'no-self-assign': 'error', 'no-setter-return': 'error', 'no-shadow-restricted-names': 'error', 'no-sparse-arrays': 'error', 'no-this-before-super': 'error', 'no-undef': 'error', 'no-unexpected-multiline': 'error', 'no-unreachable': 'error', 'no-unsafe-finally': 'error', 'no-unsafe-negation': 'error', 'no-unused-labels': 'error', 'no-unused-vars': 'error', 'no-useless-catch': 'error', 'no-useless-escape': 'error', 'no-with': 'error', 'require-yield': 'error', 'use-isnan': 'error', 'valid-typeof': 'error' }, settings: undefined }, { type: 'config', name: '.eslintrc.js', filePath: 'C:\\work\\vdm-services\\.eslintrc.js', criteria: null, env: { node: true, es2020: true }, globals: undefined, ignorePattern: IgnorePattern { patterns: 
[Array], basePath: 'C:\\work\\vdm-services', loose: false }, noInlineConfig: undefined, parser: { error: null, filePath: 'C:\\work\\vdm-services\\node_modules\\@typescript-eslint\\parser\\dist\\index.js', id: '@typescript-eslint/parser', importerName: '.eslintrc.js', importerPath: 'C:\\work\\vdm-services\\.eslintrc.js' }, parserOptions: { ecmaVersion: 2019, sourceType: 'module', project: './tsconfig.json' }, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { 'no-unused-vars': 'off', 'no-constant-condition': [Array], 'no-console': 'error', eqeqeq: 'error', 'max-classes-per-file': [Array], 'max-len': [Array], semi: [Array], 'quote-props': [Array], quotes: [Array], 'new-parens': 'error', 'no-caller': 'error', 'no-bitwise': 'error', indent: [Array], 'no-floating-decimal': 'error', 'comma-dangle': [Array], 'comma-spacing': 'error', 'no-trailing-spaces': 'error', curly: 'error', 'prefer-arrow-callback': 'error', 'object-shorthand': 'error', 'multiline-ternary': [Array], 'operator-linebreak': [Array], 'no-promise-executor-return': 'error', 'no-template-curly-in-string': 'error', 'no-throw-literal': 'error', 'no-useless-concat': 'error', 'no-useless-return': 'error', 'no-shadow': 'error', '@typescript-eslint/no-unnecessary-condition': [Array], '@typescript-eslint/lines-between-class-members': [Array], '@typescript-eslint/no-non-null-assertion': 'off', '@typescript-eslint/ban-types': [Array], '@typescript-eslint/explicit-module-boundary-types': 'off', '@typescript-eslint/no-unused-vars': 'off', '@typescript-eslint/no-explicit-any': 'off', '@typescript-eslint/no-unused-expressions': 'error', '@typescript-eslint/await-thenable': 'error', '@typescript-eslint/naming-convention': [Array], '@typescript-eslint/array-type': [Array], '@typescript-eslint/no-floating-promises': 'error', '@typescript-eslint/no-misused-promises': [Array], '@typescript-eslint/no-unnecessary-type-assertion': 'error', '@typescript-eslint/require-await': 'error', '@typescript-eslint/member-ordering': [Array], 'no-restricted-imports': [Array] }, settings: undefined }, { type: 'config', name: '.eslintrc.js#overrides[0]', filePath: 'C:\\work\\vdm-services\\.eslintrc.js', criteria: { includes: [Array], excludes: null, basePath: 'C:\\work\\vdm-services' }, env: { mocha: true }, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { 'no-unused-expressions': 'off', '@typescript-eslint/no-unused-expressions': 'off', 'no-console': 'off' }, settings: undefined }, { type: 'config', name: '.eslintrc.js#overrides[1]', filePath: 'C:\\work\\vdm-services\\.eslintrc.js', 
criteria: { includes: [Array], excludes: null, basePath: 'C:\\work\\vdm-services' }, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { 'no-restricted-imports': [Array] }, settings: undefined } ] on C:\work\vdm-services\packages\testdata\helper-scripts +456ms
  eslint:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: 'C:\\work\\vdm-services', loose: false }, IgnorePattern { patterns: [ 'packages/*/dist/**/*', 'packages/*/node_modules/**/*', 'packages/**/*.d.ts', '.eslintrc.js', 'packages/*/*.ts' ], basePath: 'C:\\work\\vdm-services', loose: false } ] +1s
  eslint:ignore-pattern   processed: { basePath: 'C:\\work\\vdm-services', patterns: [ '/**/node_modules/*', 'packages/*/dist/**/*', 'packages/*/node_modules/**/*', 'packages/**/*.d.ts', '.eslintrc.js', 'packages/*/*.ts' ] } +35ms
  eslint:ignore-pattern Check {
  filePath: 'C:\\work\\vdm-services\\packages\\testdata\\helper-scripts\\foo.ts',
  dot: false,
  relativePath: 'packages/testdata/helper-scripts/foo.ts',
  result: false
} +6ms
  eslint:cli-engine Lint C:\work\vdm-services\packages\testdata\helper-scripts\foo.ts +0ms
  eslint:linter Linting code for C:\work\vdm-services\packages\testdata\helper-scripts\foo.ts (pass 1) +0ms
  eslint:linter Verify +1ms
  eslint:linter With ConfigArray: C:\work\vdm-services\packages\testdata\helper-scripts\foo.ts +0ms
  eslint:linter Generating fixed text for C:\work\vdm-services\packages\testdata\helper-scripts\foo.ts (pass 1) +14s
  eslint:source-code-fixer Applying fixes +0ms
  eslint:source-code-fixer shouldFix parameter was false, not attempting fixes +0ms
  eslint:file-enumerator Complete iterating files: ["packages\\testdata\\helper-scripts\\foo.ts"] +15s
  eslint:cli-engine Linting complete in: 15482ms +14s

C:\work\vdm-services\packages\testdata\helper-scripts\foo.ts
  14:13  error  Unnecessary conditional, value is always falsy  @typescript-eslint/no-unnecessary-condition

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Versions**
3.10.0

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.10.0` |
| `@typescript-eslint/parser`        | `3.10.0` |
| `TypeScript`                       | `4..0.2` |
| `ESLint`                           | `7.7.0` |
| `node`                             | `12.16.3` |

**Comment by bradzacher** [comment]
> - [x] I have updated to the latest version of the packages.

We released a new version ~6 hours ago which addressed this.
